File: apps/web/lib/source-url.ts
export type SourceType = "youtube" | "twitch" | "upload";

export function detectSourceType(url: string): SourceType {
  let parsed: URL;
  try {
    parsed = new URL(url.trim());
  } catch {
    throw new Error("Enter a valid http(s) URL.");
  }

  const host = parsed.hostname.replace(/^www\./, "").toLowerCase();

  if (
    host === "youtube.com" ||
    host === "youtu.be" ||
    host === "m.youtube.com" ||
    host.endsWith(".youtube.com")
  ) {
    return "youtube";
  }

  if (
    host === "twitch.tv" ||
    host.endsWith(".twitch.tv") ||
    host === "clips.twitch.tv"
  ) {
    return "twitch";
  }

  throw new Error(
    "Unsupported URL. Use a YouTube or Twitch VOD link (file upload comes later).",
  );
}

export function defaultProjectName(url: string): string {
  try {
    const parsed = new URL(url.trim());
    const host = parsed.hostname.replace(/^www\./, "");
    const id =
      parsed.searchParams.get("v") ??
      parsed.pathname.split("/").filter(Boolean).pop();
    return id ? `${host} — ${id}` : host;
  } catch {
    return "New project";
  }
}

File: apps/worker/worker/media/source_url.py
"""Detect video source platform from a URL (mirrors apps/web/lib/source-url.ts)."""

from __future__ import annotations

from urllib.parse import urlparse


def detect_source_type(url: str) -> str:
    parsed = urlparse(url.strip())
    if parsed.scheme not in ("http", "https") or not parsed.netloc:
        raise ValueError("Enter a valid http(s) URL.")

    host = parsed.netloc.removeprefix("www.").lower()

    if (
        host == "youtube.com"
        or host == "youtu.be"
        or host == "m.youtube.com"
        or host.endswith(".youtube.com")
    ):
        return "youtube"

    if (
        host == "twitch.tv"
        or host.endswith(".twitch.tv")
        or host == "clips.twitch.tv"
    ):
        return "twitch"

    raise ValueError(
        "Unsupported URL. Use a YouTube or Twitch link for reference clip download."
    )
